Writing expressions (#5996)
qgis/QGIS-Documentation: docs/user_manual/working_with_vector/expression.rst

offer a powerful way to manipulate attribute value, geometry and variables
in order to dynamically change the geometry style, the content or position
of the label, the value for diagram, the height of a layout item,

> select some features, create virtual field...

Reviewer: ```suggestion
select some features, create virtual field, ...
```

Author: I'm trying to avoid the current html rendering...

Reviewer: > I'm trying to avoid the current html rendering...

I can't see anything wrong with the "current html rendering". Adding the comma, and keeping the space is the correct way to use ... in this case.

* :ref:`Function Editor tab <function_editor>` which helps to extend the list of
  functions by creating custom ones.

The Interface
-------------

The :guilabel:`Expression` tab provides the main interface to write expressions

> using functions, layer's fields and values. It contains following widgets:

Reviewer: ```suggestion
using functions, layer fields and values. It contains the following widgets:
```

.. _figure_expression_tab:

.. figure:: img/function_list.png
   :align: center
   :width: 100%

   The Expression tab

> * An expression editor area to type or paste expressions. Autocompletion is

Reviewer: ```suggestion
* An expression editor area for typing or pasting expressions. Autocompletion is
```

   collapsed (invisible) in the dialog. Press the :guilabel:`Show Values`
   or :guilabel:`Show Help` button to get it back.

Writing an expression
---------------------

QGIS expressions are used to select features or set values.
Writing an expression in QGIS follows some rules:

#. **The dialog defines the context**: if you are used to SQL, you probably
   know queries of the type *select features from layer where condition*
   or *update layer set field = new_value where condition*.
   A QGIS expression also needs all these information but the tool you use
   to open the expression builder dialog provides parts of them.
   For example, giving a layer (building) with a field (height):

   * pressing the |expressionSelect|:sup:`Select by expression` tool means that
     you want to "select features from buildings". The **condition** is the

>      remaining and single information you need to fill in the expression text widget,

Reviewer: ```suggestion
     only information you need to provide in the expression text widget,
```

>      eg type ``height > 20`` to select buildings of height greater than 20.

Reviewer: ```suggestion
     e.g. type ``"height" > 20`` to select buildings that are higher than than 20.
```

   * with this selection made, pressing the |calculateField| :sup:`Field calculator`
     button and choosing "height" as :guilabel:`Update existing field`, you already
     provide the command "update buildings set height = ??? where height > 20".

>      The only remaining bits you have to fill in this case is the **new value**,

Reviewer: ```suggestion
     The only remaining bits you have to provide in this case is the **new value**,
```

>      eg just enter ``50`` to raise previously selected buildings height.

Reviewer: ```suggestion
     e.g. just enter ``50`` to set the height of the previously selected buildings.
```

> #. **To the quotes you'll pay attention**: single quote returns a litteral, so

Reviewer: ```suggestion
#. **Pay attention to quotes**: single quotes returns a literal, so
```

>    placing a text between single quotes (``'mytext'``) is interpreted as string.

Reviewer: ```suggestion
   text placed between single quotes (``'145'``) is interpreted as a string.
```

   Double quotes will give you the value of that text so use it for fields
   (ie ``"myfield"``). Field can also be used without quotes (``myfield``).
   No quotes for numbers (``3.16``).

   .. note:: Functions normally take as argument a string for field name.
       Do::

>         attribute( @atlas_feature, 'height' ) -- returns value stored in "height" attribute for the current atlas feature

Reviewer: ```suggestion
        attribute( @atlas_feature, 'height' ) -- returns the value stored in the "height" attribute of the current atlas feature
```

       And not::

>         attribute( @atlas_feature, "height" ) -- fetches the value of the attribute named "height", and then gets the attribute from atlasfeature which is called that value. Probably wrong.

Reviewer: ```suggestion
        attribute( @atlas_feature, "height" ) -- fetches the value of the attribute named "height", and then gets the value of the attribute that has the name found in the height attribute of the atlas feature. Probably wrong.
```
?

.. index:: Named parameters
   single: Expressions; Named parameters
   single: Functions; Named parameters

> .. tip:: **Use named parameters to improve the expression reading**

Reviewer: ```suggestion
.. tip:: **Use named parameters to ease expression reading**
```

  Some functions require many parameters to be set. The expression engine supports the
  use of named parameters. This means that instead of writing the cryptic expression
  ``clamp( 1, 2, 9)``, you can use ``clamp( min:=1, value:=2, max:=9)``. This also allows
  arguments to be switched, e.g. ``clamp( value:=2, max:=9, min:=1)``. Using named parameters
  helps clarify what the arguments for an expression function refer to, which is helpful

>   when you are trying to interpret an expression at a later date!

Reviewer: ```suggestion
  when you are trying to interpret an expression later!
```

Some use cases of expressions
-----------------------------

> * From Field Calculator, calculate a "pop_density" field using existing "total_pop"

Reviewer: ```suggestion
* From the Field Calculator, calculate a "pop_density" field using the existing "total_pop"
```

  and "area_km2" fields::

    "total_pop" / "area_km2"

* Label or categorize features based on their area::

    CASE WHEN $area > 10 000 THEN 'Larger' ELSE 'Smaller' END

* Update the field "density_level" with categories according to the "pop_density" values::

    CASE WHEN "pop_density" < 50 THEN 'Low population density'
         WHEN "pop_density" >= 50 and "pop_density" < 150 THEN 'Medium population density'
         WHEN "pop_density" >= 150 THEN 'High population density'
    END

* Apply a categorized style to all the features according to whether their average house
  price is smaller or higher than 10000€ per square metre::

    "price_m2" > 10000

* Using the "Select By Expression..." tool, select all the features representing
  areas of “High population density” and whose average house price is higher than
  10000€ per square metre::

    "density_level" = 'High population density' and "price_m2" > 10000

>   Likewise, the previous expression could also be used to define which features
>   should be labeled or shown in the map.

Reviewer: ```suggestion
  The previous expression could also be used to define which features
  to label or show on the map.
```

> * Render a different symbol (type) for the layer, using the geometry generator::

Reviewer: ```suggestion
* Create a different symbol (type) for the layer, using the geometry generator::
```
?

Author: This is a matter of taste I think, so I opt to keep it and not reset translators work

Reviewer: Render is something that QGIS does. Rendering has to do with system-created visual output. In this case a new geometry is created using the generator. QGIS then renders this geometry according to the symbology defined.

    point_on_surface( $geometry )

> * Given a point feature, generate a closed line (using ``make_line``) around the
>   point's geometry::

Reviewer: ```suggestion
* Given a point feature, generate a closed line (using ``make_line``) around its
  geometry::
```

    make_line(
      -- using an array of points placed around the original
      array_foreach(
        -- list of angles for placing the projected points (every 90°)
        array:=generate_series( 0, 360, 90 ),
        -- translate the point 20 units in the given direction (angle)
        expression:=project( $geometry, distance:=20, azimuth:=radians( @element ) )
      )
    )

.. index:: Functions
.. _functions_list:

List of functions
=================

> Below are listed all the functions, operators and variables available in QGIS,

Reviewer: ```suggestion
The functions, operators and variables available in QGIS are listed below,
```
